src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+      const location = path.node.loc;
+      if (path.isProgram()) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Script", parent, location);
+        parent = scope;
+        return;
+      }
+      if (
+        path.isFunctionDeclaration() ||
+        path.isFunctionExpression() ||
+        path.isArrowFunctionExpression()
+      ) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Function", parent, location);
+        if (isNode(tree.id, "Identifier")) {
+          const functionName = { kind: "fn", refs: [] };
+          _getFunctionScope(parent).names[tree.id.name] = functionName;
+          scope.names[tree.id.name] = functionName;
+        }
+        tree.params.forEach(param => {
+          _parseDeclarator(param, scope, "param");
+        });
+        parent = scope;
+        return;
+      }
+      if (path.isForInStatement() || path.isForStatement()) {
+        const init = tree.init || tree.left;
+        let scope = parent;
+        if (
+          isNode(init, "VariableDeclaration") &&
+          (init.kind === "let" || init.kind === "const")
+        ) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("For", parent, location);
+          parent = scope;
+        }
+        return;
+      }
+      if (path.isCatchClause()) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Catch", parent, location);
+        _parseDeclarator(tree.param, scope, "param");
+        parent = scope;
+        return;
+      }
+      if (path.isBlockStatement()) {
+        const hasLet = tree.body.some(node => {
+          if (!isNode(node, "VariableDeclaration")) {
+            return false;
+          }
+          return node.kind === "let" || node.kind === "const";
+        });
+        let scope = parent;
+        if (hasLet) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("Block", parent, location);
+          parent = scope;
+        }
+        return;
+      }
+      if (path.isVariableDeclaration()) {
+        let hoistAt = parent;
+        if (tree.kind !== "let" && tree.kind !== "const") {
+          hoistAt = _getFunctionScope(parent);
+        }
+        tree.declarations.forEach(declarator => {
+          _parseDeclarator(declarator.id, hoistAt, tree.kind);
+        });
+        return;
+      }
+      if (path.isIdentifier()) {
+        _recordIdentifierPosition(parent, tree.name, tree.loc);
+      }
+    },
+    exit(path: NodePath) {
+      const savedParent = savedParents.get(path);
+      if (savedParent) {
+        parent = savedParent;
+        savedParents.delete(path);
+      }
+    }
+  });
+  return _toGeneratedScopes(parent.scopes, source.id);
+}
+
+function _compareLocations(a: Location, b: Location): number {
+  // $FlowIgnore
+  return a.line == b.line ? a.column - b.column : a.line - b.line;
+}
+
+var parsedScopesCache = new Map();
```

Comment:
we are using var here, is there a particular reason? i would expect let to be used in this case